prow/artifact-uploader/controller.go

```diff
@@ -0,0 +1,197 @@
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/test-infra/prow/pod-utils/decorate"
+
+	"k8s.io/test-infra/prow/gcsupload"
+	"k8s.io/test-infra/prow/kube"
+	"k8s.io/test-infra/prow/pod-utils/downwardapi"
+	"k8s.io/test-infra/prow/pod-utils/gcs"
+)
+
+const (
+	// ContainerLogDir is the prefix under which we place
+	// container logs in cloud storage
+	ContainerLogDir = "logs"
+)
+
+// item describes a container that we saw finish execution
+type item struct {
+	namespace     string
+	podName       string
+	containerName string
+	prowJobId     string
+}
+
+func NewController(client core.CoreV1Interface, prowJobClient *kube.Client, gcsConfig *gcsupload.Options) Controller {
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	optionsModifier := func(options *metav1.ListOptions) {
+		req, _ := labels.NewRequirement(kube.ProwJobIDLabel, selection.Exists, []string{})
+		options.LabelSelector = req.String()
+	}
+	podListWatcher := cache.NewFilteredListWatchFromClient(client.RESTClient(), "pods", api.NamespaceAll, optionsModifier)
+	indexer, informer := cache.NewIndexerInformer(podListWatcher, &api.Pod{}, 0, cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(old interface{}, new interface{}) {
+			oldPod := old.(*api.Pod)
+			newPod := new.(*api.Pod)
+
+			containers := findFinishedContainers(oldPod.Status.InitContainerStatuses, newPod.Status.InitContainerStatuses)
+			containers = append(containers, findFinishedContainers(oldPod.Status.ContainerStatuses, newPod.Status.ContainerStatuses)...)
+
+			for _, container := range containers {
+				queue.Add(item{
+					namespace:     newPod.Namespace,
+					podName:       newPod.Name,
+					containerName: container,
+					prowJobId:     newPod.Labels[kube.ProwJobIDLabel],
+				})
+			}
+
+		},
+	}, cache.Indexers{})
+
+	return Controller{
+		queue:         queue,
+		indexer:       indexer,
+		informer:      informer,
+		client:        client,
+		prowJobClient: prowJobClient,
+		gcsConfig:     gcsConfig,
+	}
+}
+
+func findFinishedContainers(old, new []api.ContainerStatus) []string {
+	var containerNames []string
+	for _, oldInitContainer := range old {
+		if oldInitContainer.Name == decorate.TestContainerName {
+			// logs from the test container will be uploaded by the
+			// sidecar, so we do not need to worry about them here
+			continue
+		}
+		for _, newInitContainer := range new {
+			// we need to take action if we see a container that is
+			// terminated that was not terminated last time we saw it
+			if oldInitContainer.Name == newInitContainer.Name &&
+				oldInitContainer.State.Terminated == nil && newInitContainer.State.Terminated != nil {
+				containerNames = append(containerNames, newInitContainer.Name)
+			}
+		}
+	}
+	return containerNames
+}
+
+type Controller struct {
+	queue    workqueue.RateLimitingInterface
+	indexer  cache.Indexer
+	informer cache.Controller
+
+	client        core.CoreV1Interface
+	prowJobClient *kube.Client
```

Comment:
Any need for a client vs a lister that uses cached prowjobs?

Reply:
Are you suggesting I build a lister and a cache, or do we already have one?

Reply:
(We can get a lister and cache nicely later when I refactor the ProwJobs and we generate clients for them using the CRD generators)

Reply:
Agreed